test/framework/src/main/java/org/elasticsearch/transport/nio/ESSelector.java

```diff
@@ -0,0 +1,161 @@
+/*
+ * Licensed to Elasticsearch under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.elasticsearch.transport.nio;
+
+import org.elasticsearch.common.util.BigArrays;
+import org.elasticsearch.transport.nio.channel.NioChannel;
+
+import java.io.IOException;
+import java.nio.channels.ClosedSelectorException;
+import java.nio.channels.Selector;
+import java.util.Collections;
+import java.util.Set;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ConcurrentLinkedQueue;
+import java.util.concurrent.CountDownLatch;
+
+public abstract class ESSelector {
+
+    protected static final int NOT_STARTED = 0;
+    protected static final int RUNNING = 1;
+    protected static final int STOPPED = 2;
+
+    protected Selector selector;
+    protected volatile int state = 0;
+    protected final ConcurrentLinkedQueue<NioChannel> channelsToClose = new ConcurrentLinkedQueue<>();
+    protected final Set<NioChannel> registeredChannels = Collections.newSetFromMap(new ConcurrentHashMap<NioChannel, Boolean>());
+
+    private final EventHandler eventHandler;
+    private final BigArrays bigArrays;
+    private volatile Thread thread;
+    private CountDownLatch shutdownLatch = new CountDownLatch(1);
```

Comment:
The main reason for the CountDownLatch was to ensure that caller knew that all closing steps had been completed. I modified this to keep the latch internal and to just wait inside of close(). 

That honestly seemed like the simplest and in retrospect there was no reason to leak the latch. Let me know if that seems reasonable to you.